base/broadcast.jl

```diff
@@ -1385,4 +1385,50 @@ function Base.show(io::IO, op::BroadcastFunction)
 end
 Base.show(io::IO, ::MIME"text/plain", op::BroadcastFunction) = show(io, op)
 
+struct IMArrayStyle <: Broadcast.AbstractArrayStyle{Any} end
+BroadcastStyle(::Type{<:Core.ImmutableArray}) = IMArrayStyle()
+
+#similar has to return mutable array 
+function Base.similar(bc::Broadcasted{IMArrayStyle}, ::Type{ElType}) where ElType
+    similar(Array{ElType}, axes(bc))
+end
+
+#but we define im_similar for giving a similar immutable array
+function im_similar(bc::Broadcasted{IMArrayStyle}, ::Type{ElType}) where ElType
+    Core.ImmutableArray(similar(bc, ElType))
+end
+
+#for now, make IMArrayStyle beat others to keep immutability
```

Comment:
If I read the `copy` function code below, immutability is the same as other styles, but adds a defensive copy at the end. Why do we want to keep it as the preference?

Actually, can you talk about the motivation more for `IMArrayStyle`? It seems to be merely a mimic of AbstractArrayStyle, and the user could make the `freeze` copy, if they needed it.